eruditus/lib/discord_util.py

```diff
@@ -0,0 +1,246 @@
+            )
+            return None
+
+        return ctf
+
+    # Trying to find ctf by provided name
+    ctf = MONGO[DBNAME][f"{CTF_COLLECTION}"].find_one(
+        {"name": re.compile(f"^{re.escape(name.strip())}$", re.IGNORECASE)}
+    )
+    if ctf is None:
+        await interaction.followup.send("No such CTF.")
+        return None
+
+    return ctf
+
+
+async def get_challenge_solvers(
+    interaction: discord.Interaction, challenge: dict, members: Optional[str] = None
+) -> list[str]:
+    if interaction.user.name not in challenge["players"]:
+        challenge["players"].append(interaction.user.name)
+
+    result: list[str] = [interaction.user.name]
+
+    if members is None:
+        return result
+
+    for mention in re.split(r"<@!?([0-9]{15,20})>", members):
+        if mention == "":
+            continue
+
+        # If id and we can fetch the user
+        if re.match(r"^[0-9]{15,20}$", mention):
+            member_info = await interaction.guild.fetch_member(int(mention))
+
+            if member_info:
+                result.append(member_info.name)
+                continue
+
+        # Otherwise, adding as a plain text
+        result.extend(mention.split())
+
+    return result
+
+
+async def mark_if_maxed(interaction: discord.Interaction, challenge: dict) -> None:
+    solved_states = MONGO[DBNAME][CHALLENGE_COLLECTION].aggregate(
+        [
+            {"$match": {"category": challenge["category"]}},
+            {"$project": {"_id": 0, "solved": 1}},
+        ]
+    )
+    if any(not state["solved"] for state in solved_states):
+        return
+
+    text_channel = interaction.channel.parent
+    if text_channel.name.startswith("🔄"):
+        await text_channel.edit(name=text_channel.name.replace("🔄", "🎯"))
+
+
+async def add_challenge_solver(
+    interaction: discord.Interaction, challenge: dict
+) -> discord.Thread:
+    challenge["players"].append(interaction.user.name)
+
+    challenge_thread = discord.utils.get(
+        interaction.guild.threads, id=challenge["thread"]
+    )
+
+    await challenge_thread.add_user(interaction.user)
+
+    MONGO[DBNAME][CHALLENGE_COLLECTION].update_one(
+        {"_id": challenge["_id"]},
+        {"$set": {"players": challenge["players"]}},
+    )
+
+    return challenge_thread
+
+
+async def remove_challenge_solver(
+    interaction: discord.Interaction, challenge: dict, send_msg: bool = True
+) -> discord.Thread:
+    challenge["players"].remove(interaction.user.name)
+
+    MONGO[DBNAME][CHALLENGE_COLLECTION].update_one(
+        {"_id": challenge["_id"]},
+        {"$set": {"players": challenge["players"]}},
+    )
+
+    challenge_thread = discord.utils.get(
+        interaction.guild.threads, id=challenge["thread"]
+    )
+
+    if send_msg:
+        await challenge_thread.send(
+            f"{interaction.user.mention} left you alone, what a chicken! 🐥"
+        )
+
+    await challenge_thread.remove_user(interaction.user)
+    return challenge_thread
```

Comment:
A bit confusing that this function returns something (`challenge_thread`). How about making the challenge thread an argument as well? The function signature would be something like:
```python
async def remove_challenge_solver(
    thread: discord.Thread, challenge: dict, user: discord.User, send_msg: bool = True
) -> None:
```
Also, not sure about the `solver` bit of this function name, maybe `remove_challenge_worker`?